### PR TITLE
Flash: Add onchain fees to settlement TX

### PR DESCRIPTION
**Pull request by omerfirmak**
this utilizes newly added Sighash.OmitSingle

### CODE REVIEW

File: source/agora/flash/Channel.d [hunk 1]
```diff
@@ -2270,6 +2264,56 @@ LOuter: while (1)
...
+        Prepare the settlement TX and publish
+
+        Params:
+            update = update to be published
+
+    ***************************************************************************/
+
+    protected Transaction publishSettlementTx (in UpdatePair update)
+    {
+        auto settle_tx = update.settle_tx.serializeFull.deserializeFull!Transaction();
```
Comment (hewison-chris) [inline source/agora/flash/Channel.d]:
```suggestion
        auto settle_tx = update.settle_tx.clone();
```

File: source/agora/flash/Types.d [hunk 2]
```diff
@@ -98,7 +98,10 @@ public struct UpdatePair
     public Transaction settle_tx;
 
     /// Our portion of the settlement multi-sig
-    public Signature our_settle_sig;
+    public SigPair our_settle_sig;
+
+    /// Our portion of the settlement multi-sig
```
Comment (hewison-chris) [inline source/agora/flash/Types.d]: Comment needs to be updated here

**Review by hewison-chris — CHANGES_REQUESTED** [review]
LGTM, just a couple of small requests. Also needs a rebase.

File: source/agora/flash/Channel.d [hunk 3]
```diff
@@ -2270,6 +2264,56 @@ LOuter: while (1)
...
     }
 
+    /***************************************************************************
+
+        Prepare the settlement TX and publish
+
+        Params:
+            update = update to be published
+
+    ***************************************************************************/
```
Comment (linked0) [inline source/agora/flash/Channel.d:2280]: Should add documentation for the return value.

File: source/agora/flash/Channel.d [hunk 4]
```diff
@@ -2270,6 +2264,56 @@ LOuter: while (1)
...
+        settle_tx.inputs[0].utxo = this.last_externalized_update_utxo;
+
+        auto utxos = this.getFeeUTXOs(settle_tx.sizeInBytes());
+        settle_tx.inputs ~= utxos.utxos.map!(hash => Input(hash)).array;
+        settle_tx.inputs.sort();
+
+        if (utxos.total_value.sub(utxos.total_fee) && utxos.total_value > 0.coins)
+            settle_tx.outputs[update.multi_settle_sig.output_idx] = Output(utxos.total_value, this.flash_conf.key_pair.address);
+        else
+            // todo: this is a requirement of SigHash.OmitSingle, fix the need to always have a refund output
```
Comment (linked0) [inline source/agora/flash/Channel.d:2296]: Shouldn't we create an issue for this for a better explanation about the problem?

**Review by linked0 — APPROVED** [review]
LGTM, just a couple of small requests.

File: source/agora/flash/Channel.d (hunk 1 above)
Comment (omerfirmak) [reply source/agora/flash/Channel.d]: Ah nice, didnt know that existed:)

File: source/agora/flash/Channel.d (hunk 4 above)
Comment (omerfirmak) [reply source/agora/flash/Channel.d:2296]: Ok, I will raise an issue for it,

**Comment by omerfirmak** [comment]
All comments addressed.

**Comment by omerfirmak** [comment]
@hewison-chris waiting for your approval

File: source/agora/flash/Channel.d [hunk 5]
```diff
@@ -2270,6 +2267,59 @@ LOuter: while (1)
...
+            settle_tx.outputs[update.multi_settle_sig.output_idx] = Output(Amount(1), this.flash_conf.key_pair.address);
```
Comment (hewison-chris) [inline source/agora/flash/Channel.d:2297]: I think allowing an `Output` to have an `Amount` of zero (just until the issue for `SigHash.OmitSingle` is resolved) is better than creating Coin units from nothing.

File: source/agora/flash/Channel.d (hunk 4 above)
Comment (hewison-chris) [reply source/agora/flash/Channel.d:2296]: Did you raise this issue?

**Review by hewison-chris — COMMENTED** [review]
If I understand correctly that this will increase the total coin supply then I think we should fix the need to have a fake output first.

File: source/agora/flash/Channel.d (hunk 5 above)
Comment (omerfirmak) [reply source/agora/flash/Channel.d:2297]: We are not creating any coins. that is impossible thru user TXs, slnce TX validation would catch that. We are only paying less fees, although by just a single unit.

File: source/agora/flash/Channel.d (hunk 4 above)
Comment (omerfirmak) [reply source/agora/flash/Channel.d:2296]: Just did, https://github.com/bosagora/agora/issues/2753

**Review by hewison-chris — APPROVED** [review]
LGTM, although I assume you will add some more description issue https://github.com/bosagora/agora/issues/2753.

**Pull request merged**